x/protorev/keeper/rebalance.go

```diff
@@ -167,30 +173,151 @@ func (k Keeper) FindMaxProfitForRoute(ctx sdk.Context, route RouteMetaData, rema
+		}
+
+		if pool.GetType() == poolmanagertypes.Concentrated {
+			clPools = append(clPools, route.PoolId)
+			inputMap[route.PoolId] = prevInput
+		}
+
+		prevInput = route.TokenOutDenom
+	}
+
+	// If there are concentrated liquidity pools in the route, then we may need to reduce the upper bound of the binary search.
+	updatedMax, err := k.ReduceSearchRangeIfNeeded(ctx, clPools, inputDenom, inputMap, curLeft, curRight)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	if updatedMax.LT(curRight) {
+		return curLeft, updatedMax, nil
+	}
+
+	return k.ExtendSearchRangeIfNeeded(ctx, route, inputDenom, curLeft, curRight, updatedMax)
+}
+
+// ReduceSearchRangeIfNeeded returns the max amount in that can be used for the binary search
+// respecting the max ticks moved across all concentrated liquidity pools in the route.
+func (k Keeper) ReduceSearchRangeIfNeeded(
+	ctx sdk.Context,
+	clPools []uint64,
+	inputDenom string,
+	inputMap map[uint64]string,
+	curLeft, curRight sdk.Int,
+) (sdk.Int, error) {
+	// Find the max amount in that can be used for the binary search (in terms of uosmo)
+	maxInputAmount, stepSize, err := k.MaxInputAmount(ctx)
+	if err != nil {
+		return sdk.ZeroInt(), err
+	}
+
+	// Iterate through all CL pools and determine the maximal amount of input that can be used
+	// respecting the max ticks moved.
+	for _, poolId := range clPools {
+		maxInForPool, _, err := k.concentratedLiquidityKeeper.ComputeMaxInAmtGivenMaxTicksCrossed(
+			ctx,
+			poolId,
+			inputMap[poolId],
+			types.MaxTicksMoved,
+		)
+
+		// In the case where we cannot calculate the max in amount, we short circuit the search
+		// and return an upper bound of 0. This will cause the binary search to not run.
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+
+		maxInForPoolInOsmo, err := k.ConvertProfits(ctx, maxInForPool, maxInForPool.Amount)
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+
+		if maxInForPoolInOsmo.LT(maxInputAmount) {
+			maxInputAmount = maxInForPoolInOsmo
+		}
+	}
+
+	return maxInputAmount.Quo(stepSize), nil
+}
+
 // Determine if the binary search range needs to be extended
-func (k Keeper) ExtendSearchRangeIfNeeded(ctx sdk.Context, route RouteMetaData, inputDenom string, curLeft, curRight sdk.Int) (sdk.Int, sdk.Int) {
+func (k Keeper) ExtendSearchRangeIfNeeded(
+	ctx sdk.Context,
+	route RouteMetaData,
+	inputDenom string,
+	curLeft, curRight, updatedMax sdk.Int,
+) (sdk.Int, sdk.Int, error) {
 	// Get the profit for the maximum amount in
 	_, maxInProfit, err := k.EstimateMultihopProfit(ctx, inputDenom, curRight.Mul(route.StepSize), route.Route)
 	if err != nil {
-		return curLeft, curRight
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
 	}
 
 	// If the profit for the maximum amount in is still increasing, then we can increase the range of the binary search
 	if maxInProfit.GTE(sdk.ZeroInt()) {
+		maxInPlusOne := curRight.Add(sdk.OneInt()).Mul(route.StepSize)
+		if updatedMax.LT(maxInPlusOne) {
```

Comment:
I think updatedMax is not multiplied by stepsize since the check previously in `UpdateSearchRangeIfNeeded` is 

```
	if updatedMax.LT(curRight) {
		return curLeft, updatedMax, nil
	}
```

so this check smells wrong to me